### PR TITLE
Fix Windows and onnx dtype compatibility

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
# What does this PR do?

Fixes multiple windows specific issues:
- dtype mismatch: numpy uses a default int representation that depends on os. On ubuntu/macos it uses `int64` but on windows it uses `int32`. This results in tokenizers returning input ids and attention masks in a different format than torch's default which is int64/long.
- shutil.rmtree doesn't always work on windows, there's some random failures that we can see when a tempfile is being closed (cleaned up) or when directly trying to shutil.rmtree a folder.

Also seq2seq decoder used to return past kv cache in torch format all the time.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: tests/onnxruntime/test_modeling.py [hunk 1]
```diff
@@ -186,7 +186,10 @@ def test_load_model_from_empty_cache(self):
         dirpath = os.path.join(default_cache_path, "models--" + self.TINY_ONNX_MODEL_ID.replace("/", "--"))
 
         if os.path.exists(dirpath) and os.path.isdir(dirpath):
-            shutil.rmtree(dirpath)
+            if os.name == "nt":
+                os.system(f"rmdir /S /Q {dirpath}")
+            else:
+                shutil.rmtree(dirpath)
```
Comment (tengomucho) [inline tests/onnxruntime/test_modeling.py:192]: why do you need this on windows? Also, you might want to consider isolating this code and make a function, to avoid repeating it.

File: optimum/onnxruntime/modeling_ort.py [hunk 2]
```diff
@@ -267,10 +267,14 @@ def __init__(
             **kwargs,
         )
 
-        self.inputs_names = {input_key.name: idx for idx, input_key in enumerate(model.get_inputs())}
+        # why are these not lists ?
+        self.input_names = {input_key.name: idx for idx, input_key in enumerate(model.get_inputs())}
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:271]: Changing the variable names makes the diff for this PR bigger, which makes it a bit harder to review

File: tests/onnxruntime/test_modeling.py (hunk 1 above)
Comment (IlyasMoutawwakil) [reply tests/onnxruntime/test_modeling.py:192]: it's this error https://github.com/huggingface/optimum/actions/runs/9366908110/job/25785437479#step:7:30120
some refs: https://github.com/python/cpython/issues/107408

File: optimum/onnxruntime/modeling_ort.py [hunk 3]
```diff
@@ -1131,45 +1161,34 @@ def forward(
         use_torch = isinstance(input_ids, torch.Tensor)
         self.raise_on_numpy_input_io_binding(use_torch)
 
+        if attention_mask is None:
+            if use_torch:
+                attention_mask = torch.ones_like(input_ids)
+            else:
+                attention_mask = np.ones_like(input_ids)
```
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_ort.py:1168]: but that makes the behavior different than transformers, and we do create them on the fly for the feature extraction class for example:
https://github.com/huggingface/optimum/blob/ac951ca0ae49983b1515b84342fea15c3b7ec35c/optimum/onnxruntime/modeling_ort.py#L995
I have already tried to do that and discussed it in https://github.com/huggingface/optimum/pull/1045 but the PR was merged too early with only one class updated.

[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:1168]: This does not work for batched inputs.

File: optimum/onnxruntime/modeling_ort.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_ort.py:271]: yeah the issue is that there were naming discrepancies which made little sense to me, especially between `ORTModel` and `ORTModelPart`
https://github.com/huggingface/optimum/blob/ac951ca0ae49983b1515b84342fea15c3b7ec35c/optimum/onnxruntime/modeling_ort.py#L270
https://github.com/huggingface/optimum/blob/ac951ca0ae49983b1515b84342fea15c3b7ec35c/optimum/onnxruntime/base.py#L54
is that intended ?
I think there might be a way to remove these naming changes, which is by fixing dtype in the seq2seq modeling classes instead of model parts. tell me if that makes more sense to you.

Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:271]: most likely not

File: optimum/onnxruntime/modeling_ort.py [hunk 4]
```diff
@@ -969,51 +1013,37 @@ def forward(
...
-                onnx_inputs["token_type_ids"] = token_type_ids
+            onnx_inputs = self._prepare_onnx_inputs(use_torch, **model_inputs)
+            onnx_outputs = self.model.run(None, onnx_inputs)
+            model_outputs = self._prepare_onnx_outputs(use_torch, *onnx_outputs)
 
-            outputs = self.model.run(None, onnx_inputs)
+            # why do we only return last_hidden_state? why not all outputs?
+            # that way, there will be less need for ORTModelForCustomTask in cases where
+            # we just want to extend model outputs with attentions, hidden_states, etc.
+            last_hidden_state = model_outputs["last_hidden_state"]
```
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_ort.py]: I added a todo to it ! I think it could be cool to support arbitrary i/o.

File: optimum/onnxruntime/modeling_ort.py [hunk 5]
```diff
@@ -852,6 +863,39 @@ def raise_on_numpy_input_io_binding(self, use_torch: bool):
                 " with model.use_io_binding = False, or pass torch.Tensor inputs instead."
             )
 
+    def _prepare_onnx_inputs(
```
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply optimum/onnxruntime/modeling_ort.py:865]: it only applies when needed so for performance I think there's no added overhead vs what used to be done.
where things can be optimized and i think can be viewed as the optimization oriented path, is i/o binding:
- pre-allocation of output buffers, not during forward but before that, and dynamically augmenting the size of output buffers when batch size changes.
- decoder models with i/o binding where static cache implementation can be used as output buffers to reduce the overhead of their creation.
- decoder models input/output synchronization which can be before and after the generation loop instead of each forward call (if that's possible).

File: tests/onnxruntime/test_modeling.py (hunk 1 above)
Comment (tengomucho) [reply tests/onnxruntime/test_modeling.py:192]: So if it is a workaround for a python issue, I believe it's even more important to isolate it in a function that should be well explained, so if one day we go back to this code we do not ask ourselves "why it is done this way"?

Comment (IlyasMoutawwakil) [reply tests/onnxruntime/test_modeling.py]: done !

File: optimum/onnxruntime/modeling_ort.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fxmarty) [reply optimum/onnxruntime/modeling_ort.py:865]: I kind of agree, just not sure what each ORTModelForXXX is for then

**Review by tengomucho — APPROVED** [review]

**Review by fxmarty — APPROVED** [review]

**Comment by tengomucho** [comment]
FWIW, numpy 2.0 has just been released, and int size in windows is now 64 bit! https://numpy.org/devdocs/numpy_2_0_migration_guide.html#windows-default-integer
